JSDK-2470 network handoff tests.

makarandp0 opened the pull request:
This PR adds functionality to control networking from the integration tests.
- a new module dockerProxy (client and server) to communicate with docker
- various APIs to control networking are exposed by docker proxy
- tests to demonstrate the API usage 
- network handoff tests that utilize the API.   

- small update to the structure of uploaded artifacts so that they are easier to browse
- 
TODO: 
- [x] Merge after https://github.com/twilio/twilio-video.js/pull/727

**Contributing to Twilio**

> All third-party contributors acknowledge that any contributions they provide will be made under the same open-source license that the open-source project is provided under.

- [x] I acknowledge that all my contributions will be made under the project's license.

manjeshbhargav commented on docker/dockerProxyServer.js:
```diff
@@ -0,0 +1,299 @@
+const isDocker = require('is-docker')();
+const version = 1.00;
+
+class DockerProxyServer {
```
@makarandp0 Docs?

manjeshbhargav commented on docker/dockerProxyServer.js:
```diff
@@ -0,0 +1,299 @@
+const version = 1.00;
+
+class DockerProxyServer {
+
```
@makarandp0 Nitpick: remove extra line?

manjeshbhargav commented on docker/dockerProxyServer.js:
```diff
@@ -0,0 +1,299 @@
+  }
+
+  startServer() {
+    return this.getCurrentNetworks().then((originalNetworks) => {
```
@makarandp0 aync/await?

manjeshbhargav commented on docker/dockerProxyServer.js:
```diff
@@ -0,0 +1,299 @@
+          { path: '/connectToDefaultNetwork', implfn: 'connectToDefaultNetwork' },
+          { path: '/getAllNetworks', implfn: 'getAllNetworks' },
+          { path: '/getCurrentNetworks', implfn: 'getCurrentNetworks' },
+        ].forEach((route) => {
```
@makarandp0 `({ endpoint, handeRequest }) =>`

manjeshbhargav commented on docker/dockerProxyServer.js:
```diff
@@ -0,0 +1,299 @@
+        app.use(cors());
+
+        [
+          { path: '/version', implfn: 'version' },
```
@makarandp0 Nitpick: `{ endpoint: '/foo', handleRequest: 'bar' }`

manjeshbhargav commented on docker/dockerProxyServer.js:
```diff
@@ -0,0 +1,299 @@
+          { path: '/getCurrentNetworks', implfn: 'getCurrentNetworks' },
+        ].forEach((route) => {
+          app.get(route.path, (req, res, next) => {
+            this[route.implfn](req.params).then(data => res.send(data)).catch(error => {
```
@makarandp0 async/await?

makarandp0 replied:
I will keep this for now, It's more common usage pattern, than async / wait.

manjeshbhargav commented on docker/dockerProxyServer.js:
```diff
@@ -0,0 +1,299 @@
+  // only connected network.
+  resetNetwork() {
+    return this.disconnectFromAllNetworks()
+        .then(() => this.connectToDefaultNetwork())
```
@makarandp0 Indent with only 2 spaces?

manjeshbhargav commented on docker/dockerProxyServer.js:
```diff
@@ -0,0 +1,299 @@
+  // removes all unused networks (created by this instance)
+  _pruneNetworks() {
+    const filters = encodeURIComponent(JSON.stringify({
+      'label': { [this._instanceLabel]: true }
```
@makarandp0 No quotes around object property.

manjeshbhargav commented on docker/dockerProxyServer.js:
```diff
@@ -0,0 +1,299 @@
+    });
+  }
+
+  getCurrentContainerId() {
```
@makarandp0 docs?

makarandp0 replied:
I understand docing client methods, but not sure if it's worth doc'ing these server internal methods. They are used internally by the webserver routes. I will mark them all as private for now.

makarandp0 commented on docker/dockerProxyClient.js:
```diff
@@ -0,0 +1,151 @@
+// load this file interactively with the server url.
+if (module.parent === null) {
+  console.log("DockerProxy loaded interactively");
+  if (process.argv.length !== 3) {
```
there isn't much to warrant this dependence yet. But will keep in mind.

makarandp0 commented on docker/dockerProxyClient.js:
```diff
@@ -0,0 +1,151 @@
+    'inspectCurrentContainer',
+    'resetNetwork',
+  ].map(func => {
+    return client[func]({}).then(( result ) => {
```
I feel this Promises loop, tests things well enought.

makarandp0 commented on docker/dockerProxyClient.js:
```diff
@@ -0,0 +1,151 @@
+  makeRequest(api, postdata) {
+    const url = this._serverUrl + api;
+    const thisRequest = this._requestId++;
+    return new Promise((resolve, reject) => {
```
I have updated client side code to use fetch, but unfortunately 'node-fetch' does not work for unix sockets /Users/mpatwardhan/work/twilio-video.js/test/lib/util.js

makarandp0 commented on docker/dockerProxyServer.js:
```diff
@@ -0,0 +1,303 @@
+  _makeRequest(options, postdata) {
+    const thisRequest = this._requestId++;
+    return new Promise((resolve, reject) => {
+      let clientRequest = http.request(options, (res) => {
```
@manjeshbhargav, I will fix this, but was hoping to find a way to auto-catch such issues so that they need not be part of the PR - there was a promising 'no-extra-parens' eslint directive, but unfortunately it does not catch this one.

makarandp0 commented:
Thank you for the feedback @manjeshbhargav - I have now updated the tests per your suggestions, 

Network handoff 'step' test is still  not 100 % stable - the one where we 1) connect new network, 2) disconnect old network, expect to reconnect) has been failing  - This seems like issue with the reconnection and not test. I have added instrumentation in the test to print out the signaling and media states when tests fail, and found that they fail because signaling state goes to disconnected. 

I think this is ready for another look.

makarandp0 commented on docker/dockerProxyServer.js:
```diff
@@ -0,0 +1,303 @@
+
+  _makeRequest(options, postdata) {
+    const thisRequest = this._requestId++;
+    return new Promise((resolve, reject) => {
```
I am trying to avoid adding any dependencies to this file so that it can be separated out easily into its own module when voice client is ready for using it.

manjeshbhargav replied:
@makarandp0 I think they can re-use it even with this refactored into its own module.

makarandp0 commented:
@manjeshbhargav - This is ready for another look

manjeshbhargav commented:
@makarandp0 LGTM. Let's disable the "step" test for now. You can merge after making sure that one of the travis builds are successful (all docker jobs pass).
